TrackletAlgorithm/MatchProcessor.h

```diff
@@ -0,0 +1,914 @@
+                  const int i,
+                  const MemType* const inmem, Args... args)
+  {
+    if (read_imem == i) {
+      data = inmem->read_mem(bx, read_addr);
+    }
+    read_inmem(data, bx, read_imem, read_addr, i+1, args...);
+  }
+
+  template<class DataType, class MemType, int nMEM, int NBits_Entries>
+  bool read_input_mems(BXType bx,
+                       ap_uint<nMEM>& mem_hasdata,
+                       ap_uint<NBits_Entries> nentries[nMEM],
+                       ap_uint<kNBits_MemAddr>& read_addr,
+                       // memory pointers
+                       const MemType* const mem0, const MemType* const mem1,
+                       const MemType* const mem2, const MemType* const mem3,
+                       const MemType* const mem4, const MemType* const mem5,
+                       const MemType* const mem6, const MemType* const mem7,
+                       const MemType* const mem8, const MemType* const mem9,
+                       const MemType* const mem10, const MemType* const mem11,
+                       const MemType* const mem12, const MemType* const mem13,
+                       const MemType* const mem14, const MemType* const mem15,
+                       const MemType* const mem16, const MemType* const mem17,
+                       const MemType* const mem18, const MemType* const mem19,
+                       const MemType* const mem20, const MemType* const mem21,
+                       const MemType* const mem22, const MemType* const mem23,
+                       DataType& data, int& nproj)
+  {
+#pragma HLS inline
+    if (mem_hasdata == 0) return false;
+
+    // 5 bits memory index for up to 32 input memories
+    // priority encoder
+    ap_uint<5> read_imem = __builtin_ctz(mem_hasdata);
+
+    // read the memory "read_imem" with the address "read_addr"
+    read_inmem(data, bx, read_imem, read_addr, 0,
+    //read_inmem(data, datamem, bx, read_imem, read_addr, 0,
+               mem0,mem1,mem2,mem3,mem4,mem5,mem6,mem7,
+               mem8,mem9,mem10,mem11,mem12,mem13,mem14,mem15,
+               mem16,mem17,mem18,mem19,mem20,mem21,mem22,mem23);
+
+    // Increase the read address
+    ++read_addr;
+
+    if (read_addr >= nentries[read_imem]) {
+      // All entries in the memory[read_imem] have been read out
+      // Prepare to move to the next non-empty memory
+      read_addr = 0;
+      mem_hasdata.clear(read_imem);  // set the current lowest 1 bit to 0
+      nproj++;
+    }
+
+    return true;
+    
+  } // read_input_mems
+
+
+  /////////////////////////////////////////////////////
+  // FIXME
+  // Move the following to Constants.hh?
+  // How to deal with these using enum?
+
+  // number of bits used to distinguish allstub memories for each layer
+  constexpr unsigned int nbits_allstubslayers[6]={3,2,2,2,2,2};
+  // number of bits used to distinguish VMs in one allstub block for each layer
+  constexpr unsigned int nbits_vmmelayers[6]={2,3,3,3,3,3};
```

Comment:
The VMRouterCM also uses these values (https://github.com/cms-L1TK/firmware-hls/blob/master/TrackletAlgorithm/VMRouterCM.h#L28-L29). I didn't add them to the Constants.h file, but if we're both using them, I guess it would make sense to move it to Constants.h?

Reply:
This is the solution I came up with for the MP, using what is currently in `Constants.h`.
https://github.com/cms-L1TK/firmware-hls/blob/e61eb824e8d49267ce4664af7ea551b68c3f0adc/TrackletAlgorithm/MatchProcessor.h#L818
This uses `nbitsallstubs` instead of having a separate one for layers and disks, and just uses the `N_LAYER` number as an offset to find the disks. Unless people prefer we have separate arrays, I think this keeps the code a bit smaller.